### PR TITLE
Add graphing capabilities to AIAssistant

### PR DESCRIPTION
**Pull request by vladtar**
This is achieved via a new tool called `GraphingTool` that gets python code from `ChatGPT` and executes it in a lambda sandbox.

graph1
graph2

## Security

REMINDER: All file contents are public.

- [x] I have ensured no secure credentials or sensitive information remain in code, metadata, comments, etc.
  - [x] Please verify that you double checked that .storybook/preview.js does not contain your participant access key details. 
  - [x] There are no temporary testing changes committed such as API base URLs, access tokens, print/log statements, etc.
- [x] These changes do not introduce any security risks, or any such risks have been properly mitigated.

Describe briefly what security risks you considered, why they don't apply, or how they've been mitigated.

### Testing
- [ ] MyDataHelps iOS app
- [ ] MyDataHelps Android app
- [x] [MyDataHelps website](mydatahelps.org)

### Documentation
- [ ] I have added relevant Storybook updates to this PR as well.
- [ ] If this feature requires a developer doc update, tag @CareEvolution/api-docs.

Consider "Squash and merge" as needed to keep the commit history reasonable on `main`.

## Reviewers

Assign to the appropriate reviewer(s). Minimally, a second set of eyes is needed ensure no non-public information is published. Consider also including:
- Subject-matter experts
- Style/editing reviewers
- Others requested by the content owner

## Summary by CodeRabbit

- **New Features**
	- Introduced support for image messages in the chat, allowing users to send and receive images.
	- Added a new `GraphingTool` to generate graphs from Python code and return them as images.
	- Added new tools for querying uploaded files and retrieving presigned URLs for downloads.
	- Enhanced the `AIAssistant` component with a new option to save graph images.
	- Updated story configuration for the `AIAssistant` component to include new controls for image handling.

- **Bug Fixes**
	- Improved message handling logic to ensure proper rendering of different message types, including images.

- **Documentation**
	- Updated type definitions and interfaces to reflect new message types and tools.

### CODE REVIEW

**Comment by vladtar** [comment]
Hooked this up so that you can pass in an option to the `AIAssistant` to save the generated graphs to your files. Still needs a lot of testing but putting it out there for feedback.

image

**Comment by reneefromhold** [comment]
@vladtar Can I see a demo of this?

File: src/components/container/AIAssistant/AIAssistant.tsx [hunk 1]
```diff
@@ -86,10 +87,34 @@ export default function (props: AIAssistantProps) {
...
-                        formatCode(toolName, toolInput)
-                            .then((formattedMessage) => {
-                                addToolMessage(streamEvent.run_id, "```js\n" + formattedMessage + "```");
-                            });
+                        let formattedMessage = await formatCode(toolName, toolInput)
+                        addMessage(streamEvent.run_id, "```js\n" + formattedMessage + "```", "tool");
+                    }
+
+                    if (toolName === "graphing") {
+                        addMessage(streamEvent.run_id, `data:image/png;base64,${streamEvent.data.output.content}`, "image");
```
Comment (ClaysGit) [inline src/components/container/AIAssistant/AIAssistant.tsx:95]: This shouldn't take for granted that the graphing tool has produced a viable image. If it fails right now, this is still run but streamEvent.data.output.content is undefined.
image

File: src/helpers/AIAssistant/Tools.ts [hunk 2]
```diff
@@ -262,3 +262,59 @@ export const GetEhrNewsFeedPageTool = tool(
     })
   }
 );
+
+export const GraphingTool = tool(
+  async (input): Promise<string> => {
+
+    let response = await fetch('https://kwejahcwmsyzidk5vde5uyv33a0saruz.lambda-url.us-east-1.on.aws', {
```
Comment (ClaysGit) [inline src/helpers/AIAssistant/Tools.ts:269]: The assistant can generate invalid python. Here's an example. `timedelta` doesn't exist.
```
import matplotlib.pyplot as plt
from datetime import datetime
from collections import Counter

# Survey completion dates
survey_dates = [
    "2024-10-01",
    "2024-10-01",
    "2024-10-01",
    "2024-10-01",
    "2024-10-01",
    "2024-10-01",
    "2024-10-01",
    "2024-10-01",
    "2024-10-01",
    "2024-10-01",
    "2024-10-01",
    "2024-10-01",
    "2024-10-01",
    "2024-10-01",
    "2024-10-01"
]

# Count surveys per day
survey_count = Counter(survey_dates)

# Generate dates for the last 2 weeks
date_range = [datetime(2024, 9, 20) + timedelta(days=i) for i in range(14)]
date_labels = [date.strftime("%Y-%m-%d") for date in date_range]

# Prepare data for plotting
counts = [survey_count.get(date, 0) for date in date_labels]

# Plotting
plt.figure(figsize=(10, 5))
plt.bar(date_labels, counts, color='skyblue')
plt.xticks(rotation=45, ha='right')
plt.xlabel('Date')
plt.ylabel('Number of Surveys Completed')
plt.title('Surveys Completed Each Day Over the Last 2 Weeks')
plt.tight_layout()
plt.show()
```

500 Response:
```
{"error": "Error: name 'timedelta' is not defined"}
```

I THINK it needed to import this from the datetime lib (I'm not a Python person). e.g. `from datetime import datetime, timedelta`

Not to over-fixate on this specific failure...our code should account for mistakes from the AI Assistant. Kyle's relevant work: we check the result of Export Explorer Copilot's sql and automatically share errors back to the assistant and ask it to try again. And in any case, this tool needs an appropriate response when the fetch fails for any reason.

File: src/helpers/AIAssistant/Tools.ts [hunk 3]
```diff
@@ -262,3 +262,59 @@ export const GetEhrNewsFeedPageTool = tool(
...
+      method: 'POST',
+      body: JSON.stringify({ code: input.code }),
+      headers: { 'Content-Type': 'application/json' }
+    }).then(response => response.json());
+
+    return response.image;
```
Comment (ClaysGit) [inline src/helpers/AIAssistant/Tools.ts:275]: This should always return a string. If something goes wrong with the response, an invalid content value can get stuck in the OpenAI message history. Subsequent requests are then rejected by OpenAI.

Maybe:
```
return response.image || "";
//or get crazy?
return response.image || "An unexpected error occurred while executing the tool; please try again.";
```

**Review by ClaysGit — CHANGES_REQUESTED** [review]
Very cool, but ran into some issues playing with it.

**Review by reneefromhold — COMMENTED** [review]
This does not work with Garmin data. Is that intentional? If yes, can that be noted, or have the default query use a unified data provider? In my test case, I confirmed across other components that Garmin was connected in Live mode but for some reason this component's stories could not access Garmin data. Is that AI Chat specific?

I realize this is outside the scope of this issue, but can a prompt property be added to the 
chat?  When scrolling through the Docs page the arrow/chat send icon is hidden behind storybook's showCode button. It's not obvious that this component is a command line without drilling into the other components.

Re appendTools, the description of the parameter is not clear. It would be helpful to know what the default tools the tools are being appended to are.

In general it would be cool to either
1) have some demo data that doesn't require an .env file for ppt data to play with 
or 
2) Associate a storybook play function to demo what to config and how to interact with the prompt

**Comment by ClaysGit** [comment]
RE: Garmin, we only give options in the Type field for v2 queries (Fitbit and Apple only currently). So the LLM just makes up a non-existent Type some  portion of the time. Furthermore, the shape of data is not always clear. Even if you tell the LLM that Garmin has a "Daily" Type, it still doesn't know that to get step data it can query for "Daily" dpps and look at `result[n].properties["steps"]` for the step data.

(this is also an issue for the web view step copilot when making survey steps that load device data)

**Comment by vladtar** [comment]
@coderabbitai pause

File: src/helpers/AIAssistant/Tools.ts [hunk 4]
```diff
@@ -262,3 +262,62 @@ export const GetEhrNewsFeedPageTool = tool(
...
+    schema: z.object({
+      category: z.string().optional().describe("Category of files to query for.")
+    })
+  }
+);
+
+export const GetUploadedFileTool = tool(
+  async (input): Promise<string> => {
+
+    let response = await MyDataHelps.getFileDownloadUrl(input.key);
```
Comment (ClaysGit) [inline src/helpers/AIAssistant/Tools.ts:311]: Any error handling or default response if the key is invalid/hallucinated?

**Review by ClaysGit — COMMENTED** [review]
One thing I notice in testing the save/retrieve image functionality is that it doesn't really have a sense for what is in the image. If I ask it `what graph did we make recently?`, it can load useful results because modified time is part of the query response. But if I ask `show me the graph with # of drinks consumed each day`, it has no idea which image that is. 

Here's an example response from the file query:
```
        {
            "participantID": "208fdcd4-59aa-4c9c-9fba-adc564ff6441",
            "participantIdentifier": "LCRCTDev-5587-2252",
            "key": "208fdcd4-59aa-4c9c-9fba-adc564ff6441/ai-graph/2024-10-08T17:33:08.346Z_graph-1728408788124.png",
            "category": "ai-graph",
            "fileName": "graph-1728408788124.png",
            "lastModified": "2024-10-08T13:33:09-04:00"
        }
```

Looking at this, it seems kinda limited in how it can query for files. I think it'd help a lot if the GraphingTool had another field in it's schema for "suggested file name" or something. Some kind of label for the graph, which it can be uploaded with, and which will help future queries identify the correct image.

**Comment by vladtar** [comment]
This is ready for a final look.

#### Saving a graph to your files:

save to files

#### Retrieving it later:

show the file

File: src/helpers/AIAssistant/Tools.ts [hunk 5]
```diff
@@ -275,6 +276,11 @@ export const GraphingTool = tool(
       }
     }).then(response => response.json());
```
[the remark this answers is not in the document]
Comment (vladtar) [reply src/helpers/AIAssistant/Tools.ts:277]: For the most part the lambda should respond with valid json even if running the code fails, there are try catches in the lambda itself. Plus if this errors the llm recovers reasonably graceful not sure a try catch would add much here.

**Comment by vladtar** [comment]
Yes, the "Save the graph" capability is a good suggestion I think I'm going to try to tackle that next. It will be interesting when to show the suggestions, perhaps some the first time the user starts interacting, then perhaps some different ones again after certain actions such as drawing a graph.

**Pull request merged**